### DISCUSSION TITLE
Generate "business" statistics (in logs)

### DISCUSSION THREAD

**Discussion by szpak** [renovatebot/renovate#23105] [Request Help] [answered]
### How are you running Renovate?

Self-hosted

### If you're self-hosting Renovate, tell us what version of Renovate you run.

35.40.0

### If you're self-hosting Renovate, select which platform you are using.

GitLab self-hosted

### Was this something which used to work for you, and then stopped?

I never saw this working

### Wanted end result.

Introducing Renovate in an organization, I started with a few testing project, added some more (still defined explicitly) and, in the end, I am using an autodiscovery feature, just adding the Renovate user to another groups in GitLab. It works pretty well, however, I am also a person who like to see a progress of own actions.

It could be useful to have Renovate(-cli) to produce some "business" stats expressing "metrics" such as number of: processed projects (in general, in onboarding, effectively active), ignored projects, new generated MRs, updated MRs, avg/max execution time per project, ...

They could be gathered (somehow) from logs and collected (in external way - not in the scope if this feature idea) to be able to generate some incremental analysis (e.g. charts in Grafana).

### What you tried so far.

I have seen that Renovate supports exporting traces in OTel format (which is great!), however, I'm not sure, if it would be possible to obtain the "metrics", I mentioned (rather to trace where processing takes the most). Exposing the metrics endpoint (https://github.com/mend/renovate-on-prem/issues/221) could be useful for the on-prem version (to some degree), but it is somehow problematic if the cli is used from a CI server.

Currently, I gather some stats by parsing logs with `jq` (e.g. top X repos by processing time or number of requests), but it requires an extra effort (and jq stills :) ) to parse JSON logs (and some of the aforementioned metrics could be harder to get). Having renovate logging them in one JSON info entry would make it much easier to get and process. WDYT?

**Comment by secustor** [comment]
Your are probably interested in https://github.com/renovatebot/renovate/issues/17661
and https://docs.renovatebot.com/examples/opentelemetry/

**Reply by szpak** [reply]
Thanks for your answer. In my question, I even mentioned OTel and its traces, but I didn't know that "metrics" are already also available. Switching to the on-prem version is some idea (assuming those metrics - e.g. current number of project "in onboarding" or "ingored" would be possible to get there.
For simple CLI execution, it would be enough to just have a log with those "current" (per execution) values printed on INFO (as a JSON entry). Having those values already for metrics would make that task much easier (for someone who knows how to get them).

**Reply by HonkingGoose** [reply]
You may also want to look at the [Swissquote user story](https://docs.renovatebot.com/user-stories/swissquote/#some-stats-after-two-years-with-renovate), it has a screenshot of their Renovate metrics dashboard. Their dashboard shows:

- Number of Renovated repositories
- Duration of Renovate runs in seconds, in four percentile "buckets"
- Time in the Queue (webhook), in four percentile "buckets"
- Number of job runs
- Webhook successes and queued repositories
- Renovate PRs in the queue
- Renovate PRs in progress
- Min queue size
- Graph of execution times

@onigoetz is there anything you want, and are allowed, to share about your company's dashboard?

**Reply by onigoetz** [reply]
Hi @HonkingGoose, yeah, sure, I can share some more info.

Here is a quick diagram of our scheduler : 
![image](https://github.com/renovatebot/renovate/assets/309594/d0025f63-da81-435a-b756-056313be0b7a)

We use GitHub's webhooks to populate the queue on relevant modifications and send an event to Influx when we're done.
`RENOVATE_WEBHOOK duration= queued=true`
When adding to the queue, we store the current time -> when picking an item, we can know how long it spent in the queue

Apart from webhooks, every hour we list all jobs and queue them.

We run ~10 jobs in parallel. Each job is a separate container started for the occasion.
From each job, we listen to the log and search for specific strings to count the create/update/merge/close, once it is done we send a measurement with these numbers, along with the status, duration, and time in queue.
`RENOVATE_RUN status=success|failure|timeout repository= duration= created= merged= closed= updated= queueDuration`

Every 5 minutes, we send a status of the queue size and number of jobs currently running.
`RENOVATE_QUEUE running= size=`

Now this is how we calculate each metric with these measurements (pseudo syntax):

* Number of Renovated repositories: `max(RENOVATE_QUEUE)` (accurate because we send a queue update right after filling the queue every hour and we deduplicate repositories if they're already in the queue)
* Duration of Renovate runs in seconds, in four percentile "buckets": `percentile(RENOVATE_RUN.duration, 90)`
* Time in the Queue (webhook), in four percentile "buckets": `percentile(RENOVATE_RUN.queueDuration, 90)`
* Number of job runs: `count(RENOVATE_RUN)`
* Webhook successes and queued repositories: `RENOVATE_WEBHOOK queued=true`
* Renovate PRs in the queue: (this one is actually jobs in the queue) `last(RENOVATE_QUEUE.size)`
* Renovate PRs in progress: (this one is actually jobs running right now) `last(RENOVATE_QUEUE.running)`
* Min queue size: `min(RENOVATE_QUEUE)`
* Graph of execution times `RENOVATE_RUN`

And some metrics we've added after the article and thus aren't in the screenshot: the number of create/update/merge/delete over time, totals, and graphed values. Very useful to see hot spots of activity

Except for "update" events, we found proper strings to search for and are confident in our detection for the count of created/merged/closed PRs. 

All in all, I would agree with the above statement that having a block of JSON we can search for at the end of the run with this information pre-formatted for us would make this cleaner to process. But we're processing the log stream anyway to write it to disk, so reading it while ongoing isn't a big deal.

I can give more details and answer more questions but can't share any code. I typed all of this from memory so there might be some inconsistencies

**Reply by HonkingGoose** [reply]
Wow! Thank you for the detailed information @onigoetz! ❤️

If I'm reading your comment right you're searching for specific strings in Renovate's log output? But the Renovate maintainers and contributors can change those strings at any time. Then your  Dashboard is missing stuff, maybe without you noticing.

So getting output from Renovate in a _known_ format, and with _proper tests_ on our end would probably help. We should also define the output and what we promise in that output. So we don't accidentally change the "contract" that users are building their dashboards on. As usual, I'll let the maintainers decide what to do here. 🙂

**Reply by onigoetz** [reply]
Yes, it's definitely a guessing game that involves a bit of source reading and log reading.
It's only informative though, missing information would be annoying but we can work with that.

A proper contract to extract information would be great though :)

**Reply by szpak** [reply]
@onigoetz Thanks for your detailed comment!

> So getting output from Renovate in a known format, and with proper tests on our end would probably help. We should also define the output and what we promise in that output. 

I think, we all three feel the same. Parsing loose strings is error prone and gently speaking "not very elegant".

I believe, all my parsed text to get some data are the subset of the already mentioned in story about Swissquote. @onigoetz What do you think about preparing an initial JSON record/entry will all the values useful/required for you, as a draft (and incentive) for the Renovate maintainers to consider it as a something to implement? Maybe (most likely?) all those values are already kept somewhere in the Renovate and maybe it would be enough to just "print" them?

**Reply by szpak** [reply]
Looking back at your case, probably there could be one general entry for numbers in general (e.g. activated, in onboarding, ingored, ... + and maybe some general timing?) and some other(s?) for timing per repo.

**Reply by onigoetz** [reply]
Sure, I can have a look at what could be interesting.

I'm not sure how far I should go with this, because I can imagine a lot of different metrics to extract by what I know of Renovate's internals
* Relative to the run 
  * various timings already collected by Renovate -> probably sent over by OTEL already
  * Number of PR created / updated / merged / closed
* Global
  * Number of PRs currently open
  * Number of PRs awaiting approval / schedule
  * Number of PRs that maintainers have refused
  * Number of Issues open
  * onboarded true / false

But for my case the only ones I'm interested in is the number of PRs changed within a run, I don't need the rest.
I don't want to add the burden to the maintainers to have these metrics if nobody's going to use them.

**Reply by rarkins** [reply]
I think it would be good to have a simple log at the end of each repo which captures:
- per run metrics (e.g. commits and PRs created/updated/merged/deleted/etc)
- per repo metrics (PRs open, awaiting, merged, etc.)

**Reply by szpak** [reply]
> I think it would be good to have a simple log at the end of each repo which captures:

Yes, and preferably also some global stats at the end of the cli execution (e.g. about repositories - total  discovered, active, in onboarding, ignored, ... + maybe aggregated stats about PRs to do not have to calculate it with `jq`).

**Reply by HonkingGoose** [reply]
> And some metrics we've added after the article and thus aren't in the screenshot: the number of create/update/merge/delete over time, totals, and graphed values. Very useful to see hot spots of activity

By the way, @onigoetz do you want to open a PR to update the screenshot? It would be nice for me, and the Renovate docs readers to see your improved dashboard.

Link to the [`.png` file for the Swissquote dashboard](https://github.com/renovatebot/renovate/blob/main/docs/usage/assets/images/swissquote_stats.png).

**Reply by HonkingGoose** [reply]
@rarkins said:

> I think it would be good to have a simple log at the end of each repo which captures:
> 
>     * per run metrics (e.g. commits and PRs created/updated/merged/deleted/etc)
> 
>     * per repo metrics (PRs open, awaiting, merged, etc.)

Issue #17661 asks for _Prometheus Metrics for Observability_. That issue is still in `status:requirements`. 

I suspect putting some metrics in a console log statement is easier than the Prometheus Metrics. 😄 I can't find any issue for "simple log at end of repo run". So, @rarkins do you want me to create a new "simple logging" issue for us?

**Reply by rarkins** [reply]
Yes please

**Reply by HonkingGoose** [reply, marked as answer]
Here's your issue! 😉 

- #23961

**Comment by HonkingGoose** [comment]
By the way, @onigoetz do you want to open a PR to add your diagram and explanations into the Renovate swissquote article? Also, the metrics under the new dashboard image don't match the image anymore.

- https://github.com/renovatebot/renovate/discussions/23105#discussioncomment-6366621
- [Swissquote article source file](https://github.com/renovatebot/renovate/blob/main/docs/usage/user-stories/swissquote.md)

**Reply by onigoetz** [reply]
Hi @HonkingGoose yeah sure, I could.

I just wasn't sure if you would be big fans of this since it diverges from your own offering, but sure, I can make a PR to add it

**Reply by HonkingGoose** [reply]
Thanks for the quick response! :heart: I'm a fan of adding the extra information to our docs. So I'm happy to accept a PR that adds the diagram and extra information, and the updated metrics.

**Reply by HonkingGoose** [reply]
@MShekow built a custom Renovate dashboard. Here's their comment from another discussion: [^1]

> By the way, I built a prototype for visualizing Renovate PRs, feel free to check it out: https://github.com/MShekow/renovate-pr-visualization

[^1]: https://github.com/renovatebot/renovate/discussions/26024#discussioncomment-7897831
